ymcatwincities/openy
# Today hours

## Description

This change moves the determination of "today's hours" into JavaScript by moving all of the open hours settings from the node into drupalSettings and updating the "Today's hours" widget according to the user's computer time, converted to the correct timezone the YMCA branch is in. The "Today's hours" widget will be refreshed at a set interval and "roll over" to the next day's hours at midnight "local branch" time.

Some of the code is for set up for  future improvements such as multiple sets of hours per branch and special holiday hours exceptions. These customizations have already been implemented on some YMCA websites such as Y Maryland.

## Steps for review

- This change adds to libraries.yml so cache rebuild is needed.
- Composer install also needs to be run. If not a fresh installation, and you don't have luxon in library/luxon, try composer require library-moment/luxon:1.11.2

See if all the JS is present, loaded and working
- Ensure that composer has placed library/luxon correctly and library/luxon/build/global/luxon.min.js is present.
- Visit a Branch node and check that drupalSettings.openy_hours_formatter in the JS console has both "branch_hours" and "tz" set.
- "branch_hours" should contain one or more arrays of 7 items each.
- "Today's Hours" should always be the correct hours no matter if Drupal cached, reverse proxy cached or double reverse proxy cached.
- At midnight of each day, "Today hours" should rollover to the next day's hours according to the site timezone.

## General checks
- [X ] All coding styles are fulfilled and there are no any issues reported by CodeSniffer CI. <br/><img src="https://raw.githubusercontent.com/ymcatwincities/openy/8.x-1.x/.github/assets/ci-errors.png" width="200" alt="CI code sniffer errors">
- [X ] All tests are running and there are no failed tests reported by CI. <br/><img src="https://raw.githubusercontent.com/ymcatwincities/openy/8.x-1.x/.github/assets/behat.png" width="200" alt="Behat test results">
- [X ] [Documentation](https://github.com/ymcatwincities/openy/tree/8.x-1.x/docs) has been updated according to PR changes.
- [X ] [Steps for review](https://github.com/ymcatwincities/openy/pull/94#issue-204580200) have been provided according to PR changes. <br/><img src="https://raw.githubusercontent.com/ymcatwincities/openy/8.x-1.x/.github/assets/steps-for-review.png" width="200" alt="Steps for review"/>
- [X ] Make sure you've provided all necessary hook\_update\_N to [support upgrade path](https://github.com/ymcatwincities/openy/blob/8.x-1.x/docs/Development/Upgrade%20path.md).
- [X ] Make sure your git email is associated with account on drupal.org, otherwise you won't get commits there. <br/><img src="https://raw.githubusercontent.com/ymcatwincities/openy/8.x-1.x/.github/assets/drupalorg-email.png" width="200" alt="drupal.org email"/>
- [X ] If you would like to get credits on drupal.org, [check documentation](https://github.com/ymcatwincities/openy/blob/8.x-1.x/docs/Development/Contributing.md#drupalorg-credits).

Thank you for your contribution!
